File: DSA/src/Questions/BinarySearch/SplitArrayLargestSum.java
package Questions.BinarySearch;

import java.util.Arrays;
/*
 https://leetcode.com/problems/split-array-largest-sum/description/
 Time Complexity: O(N * log(sum(nums))) where N is the length of the nums array. This is because we perform binary search on the range [max(nums), sum(nums)] and for each mid-point, we check the feasibility in O(N).
Space Complexity: O(1) as we use a constant amount of extra space.

 */
public class SplitArrayLargestSum {

    // Method to count the number of students needed to read the given number of
    // pages
    public static int countStudents(int[] nums, int pages) {
        int students = 1;

        long sumPages = 0;
        // Loop through the array and add the pages read by each student
        for (int i = 0; i < nums.length; i++) {
            if (sumPages + nums[i] <= pages) {
                sumPages += nums[i];
            } else {
                // If the current student cannot read more pages, increment the student count
                // and assign the current pages to the new student
                students++;
                sumPages = nums[i];
            }
        }
        return students;
    }

    // Method to split the array into m sub-arrays such that the maximum sum of the
    // sub-arrays is minimized
    public static int splitArray(int[] nums, int m) {
        int low = 0, high = 0;
        // Calculate the lower and upper bounds for binary search
        // for (int num : nums) {
        //     low = Math.max(low, num); // The maximum element in the array
        //     high += num; // The sum of all elements in the array
        // }
        low = Arrays.stream(nums).max().getAsInt();
        high = Arrays.stream(nums).sum();
        System.out.println("Low: " + low + " High: " + high);
        // Apply binary search
        while (low <= high) {
            int mid = (low + high) / 2;
            int students = countStudents(nums, mid);
            // If the number of students needed is less than or equal to m, search in the
            // lower half
            if (students <= m) {
                high = mid - 1;
            } else {
                // Else, search in the upper half
                low = mid + 1;
            }
        }
        return low;
    }

    /*
     * Time Complexity: O(N * log(sum of array elements))
     */

    // Main method
    public static void main(String[] args) {
        int[] nums = { 10, 20, 30, 40 };
        int m = 2;
        // Find the minimum possible maximum sum of m sub-arrays
        System.out.println(splitArray(nums, m));
    }
}

File: DSA/Python/Questions/BinarySearch/SplitArrayLargestSum.py
"""
# Question: Split Array Largest Sum
# Link: https://leetcode.com/problems/split-array-largest-sum/

# Time Complexity: O(n * log(sum(nums)))
# Space Complexity: O(1)

# Algorithm:
# 1. Binary search on sum range
# 2. Check if split is possible
# 3. Find minimum largest sum
"""


class SplitArrayLargestSum:
    def split_array(self, nums: list[int], m: int) -> int:
        left = max(nums)
        right = sum(nums)

        while left < right:
            mid = (left + right) >> 1

            if self.can_split(nums, m, mid):
                right = mid
            else:
                left = mid + 1

        return left

    def can_split(self, nums: list[int], m: int, target: int) -> bool:
        count = 1
        current_sum = 0

        for num in nums:
            current_sum += num
            if current_sum > target:
                count += 1
                current_sum = num
                if count > m:
                    return False

        return True


def main():
    solution = SplitArrayLargestSum()
    nums = [7, 2, 5, 10, 8]
    m = 2
    print(solution.split_array(nums, m))  # Expected: 18


if __name__ == "__main__":
    main()
